### ISSUE TITLE
Query block: Toolbar control revisions

### ISSUE THREAD

**Issue by mapk** [WordPress/gutenberg#25198]
There have been some recent additions to the toolbar controls for the Query block. Before we go too far, let's evaluate what we have currently.

![Screen Shot 2020-09-09 at 11 19 11 AM](https://user-images.githubusercontent.com/617986/92641321-682b3d80-f293-11ea-8c45-0969b4d47f98.png)

**Posts per page**
This is an important setting, however, will the Query block every produce results that include anything other than posts? My suggestion is to change this to say, "Items per page" and make sure it is in Sentence case, not Title case.

**Number of Pages**
I think this one can change to a toggle that says, "Show more than one page". The reason being is that the current "number of pages" doesn't allow a default that just shows however many pages are required to display all the posts. And it seems likely that users will want to only show ONE page with a few results, or show ALL pages with a set list of results per page. Of course toggling this "on" would add the Query Pagination block automatically if needed.

**Offset**
This is good to keep, but needs some explanation (helper text). At first I wasn't sure what this did. Can we add some text below that offers a tip? "Offsetting a list to 3 will begin the results at the 3rd item instead of the 1st."

**Categories & Tags**
I like that these are added for the basic Query block, but I imagine they will be hidden depending on the template for which they're being used. Like if this block is used on a `category-$slug.php` page, I would assume the Tag settings would not be available. Nor would the Category settings be available because the template would pass the parameters directly to the block. Am I thinking about this correctly?

**Add "Order by" sorting**
Let's add the "Order by" sorting option to this list as well.

### New mockup
With these design changes, the toolbar popover should look something like this:

![Popover](https://user-images.githubusercontent.com/617986/92643121-1afc9b00-f296-11ea-8875-956d34c44df9.png)

cc @ntsekouras

**Comment by mapk** [comment]
### Order by
I realize the "Order by" was added here: https://github.com/WordPress/gutenberg/pull/24691 to the Inspector, but I thought it might work well in this popover. Now, I'm second guessing this one because it doesn't feel like an important setting required for the use of this block. Chances are, this particular setting won't get a lot of use as most people will prefer the default of "Newest to oldest".

### Filter by Author
The ability to filter by "Author" was added here: https://github.com/WordPress/gutenberg/pull/25149. Regarding my comment above about Categories and Tags, should the Author option also hide depending on the template being used? For example, if I'm adding this block to a `author-$id.php` template, I probably don't need this setting, right?

**Comment by ntsekouras** [comment]
Hey @mapk! I don't have many answers myself yet, but here are some of my thoughts :)

In the first design iteration I think we should probably change the `items per page` and `offset` to `number` controls and not `range`. I can't think of any reason for them to be `range`. If anyone does, please comment.

**Number of Pages**
>Of course toggling this "on" would add the Query Pagination block automatically if needed.

That seems right, probably with different wording, like `show pagination`.

**Offset**
An explaining text seems a good idea to me.

**Categories & Tags**
>Am I thinking about this correctly?

I would like to hear some thoughts from people more involved in FSE as well.

**Order by**
This seems to be a `secondary` option for inspector controls.

**Filter by Author**
Even if this is conditionally shown, we should decide where to show it. It seems to me a bit more of inspector controls item.

**Comment by ntsekouras** [comment]
Also keep in mind that another `control` will be added for sure that is the `search` control which require a text input. This will be probably needed for any context.

**Comment by mapk** [comment]
>In the first design iteration I think we should probably change the items per page and offset to number controls and not range. I can't think of any reason for them to be range. If anyone does, please comment.

That's right! I remember you mentioning this before. I'm on board with this.

>That seems right, probably with different wording, like show pagination.

Maybe. I think whatever we use for copy will need to be flexible enough that if there's only enough results to show in one page, it still makes sense. I'll add the "Needs copy" label to get some thoughts.

Also cc-ing @mtias to your input on this. We're iterating on the settings for the Query block and can use some feedback regarding this particular issue.

**Comment by mapk** [comment]
Here's a mockup including some of these changes.

- I noticed labels, if possible, have been moved inline with their inputs, so I did this with the Number inputs.
- Updated the label text for each one.
- The last label, I feel, communicates the toggle really well now.
- Removed the Order by option to keep in the sidebar because it's just as important.

![block-toolbar](https://user-images.githubusercontent.com/617986/92767079-9ff5bc00-f34b-11ea-9651-fbc1731076f4.png)

**Comment by mapk** [comment]
I received some recent feedback that has encouraged me to explore a basic Query block with all the settings, whether in the toolbar or sidebar, first. From there we can explore the variations that may hide certain settings, or show settings preconfigured (and possibly uneditable if inherited from template).

There was also some thoughts on whether or not `Offset` is important enough to be in the toolbar dropdown.

_will have another iteration up today_

**Comment by carolynannewells** [comment]
Hello! Popping in here from editorial. I like what you are doing here, you are making things much clearer, especially with the explaining text. I would perhaps change the "Show more than one page" toggle to "Show multiple pages."

**Comment by mapk** [comment]
Thanks, @carolynannewells! I actually iterated a bit more earlier today and landed on, "Allow pagination of results" What do you think about that one? I'm totally willing to trust your expertise here! 

I've also made some adjustments as another iteration. This is isn't quite final. Just wanted to share my progress from today.

### Option 1 - Two setting buttons in toolbar

**Notes**
- I split the "Filters" out into its own icon button in the toolbar.
- Added the major filters there (ie. Categories, Tags, Taxonomies, Author)
- Removed the "Offset" setting from the display popover.
- Edited some of the toggle text as mentioned above.

![two-buttons](https://user-images.githubusercontent.com/617986/92828060-76f31c80-f387-11ea-80e2-490e81e4dd1c.png)


### Option 2 - Inline filter input fields
Riffing off of the work in the new [Link UI](https://github.com/WordPress/gutenberg/pull/24021), I also wanted to share some explorations I did around the filters. Hat tip to @MichaelArestad for suggesting this. I pulled the filters out into their own icon buttons in the toolbar. When clicked, they open up just like the link UI does. I'm not confident about this direction, but thought it could spur more feedback.

**Prototype**
![inline-icons](https://user-images.githubusercontent.com/617986/92828754-3fd13b00-f388-11ea-9d30-9d622862ac28.gif)

**Annotated**
![Screen Shot 2020-09-10 at 5 08 53 PM](https://user-images.githubusercontent.com/617986/92828823-54153800-f388-11ea-8ed7-c9fe8d4aedc2.png)

**Mockups**
![Screen Shot 2020-09-10 at 5 09 22 PM](https://user-images.githubusercontent.com/617986/92828877-655e4480-f388-11ea-9f62-a1646004f8eb.png)

**Mockups with a "Is" or "Is not" selection box.**
![Screen Shot 2020-09-10 at 4 21 58 PM](https://user-images.githubusercontent.com/617986/92828921-7313ca00-f388-11ea-8c58-cc40495f14aa.png)

**Comment by carolynannewells** [comment]
I think that "Allow pagination of results" works! :)

**Comment by scruffian** [comment]
Could "categories" and "tags" be dropdowns rather than inputs?

**Comment by mapk** [comment]
>Could "categories" and "tags" be dropdowns rather than inputs?

Yes! 😄  I think something like what is in place currently would be great. I'm going to refine those controls further.

![Screen Shot 2020-09-11 at 2 50 18 PM](https://user-images.githubusercontent.com/617986/92975915-386c6900-f43e-11ea-8948-d9c961ec0489.png)

**Comment by mapk** [comment]
I've been working through more iterations today.

- Display (number of items) is an input like the [Link UI](https://github.com/WordPress/gutenberg/pull/24021).
- Pagination is a togglable icon button in the toolbar.
- Categories and tags remain within a Filter icon button. 
    - Categories inherits the design from what is currently in the sidebar.
- Filtering by author or taxonomies remain in the sidebar. I'm not sure these should be broken out from Categories and Tags.

**Prototype**

![prototype](https://user-images.githubusercontent.com/617986/92978372-72d90480-f444-11ea-8a20-da4aa24f17bb.gif)

**Comment by scruffian** [comment]
> Filtering by author or taxonomies remain in the sidebar. I'm not sure these should be broken out from Categories and Tags.

Does it make sense to have the category and tag filters in the sidebar as well as in the toolbar?

**Comment by bph** [comment]
Seems the idea "Bring important settings" into the toolbar, and add the rest of the settings in sidebar, is pushed quite far. The downside is an increase in cognitive load. On top of explaining the various settings, you also ave to explain the icons, (display filters, authors, numbers, tags) too. 

They are not as clean as they appear in the mock-ups. 

The sidebar is very well established on providing settings for more complex blocks (cover, image, gallery, latest posts). 
<img width="1399" alt="Screen Shot 2020-09-14 at 10 36 34 AM" src="https://user-images.githubusercontent.com/39980/93103379-8cfc2800-f67a-11ea-9be7-43a160d798ad.png">
_Screenshot in Guteneberg  9.0 RC_ 
I felt the overlapping drop-down from the toolbar quite disconcerting and very distracting from the task at hand.

**Comment by bph** [comment]
There are a few assumptions you could make instead of trying to squeeze things into the toolbar:  

- Newest to oldest
- use the `WPAdmin > Settings > Reading > Blog pages show at most` number as default for how many posts in the list 
- set pagination to "yes" or true

**Comment by mapk** [comment]
>Seems the idea "Bring important settings" into the toolbar, and add the rest of the settings in sidebar, is pushed quite far.

I think these explorations were important to visually work through all of this. You bring up a good point about cognitive load. The toolbar settings work well as dropdown select items, but when we add more form-link inputs whose input value isn't reflected in the toolbar, it gets a bit awkward.

I'm going to push in the other direction (back to the sidebar) for some of this to share.

>set pagination to "yes" or true

I was thinking this same thing too! Most likely, the Query block will be used _with_ pagination, so making that the default sounds reasonable.

**Comment by ntsekouras** [comment]
>I felt the overlapping drop-down from the toolbar quite disconcerting and very distracting from the task at hand.

For now it is, but this is the time and the issue where we will change this :). There have been some ongoing development efforts to enhance the functionality of the `Query` block, but there is always in mind that the design is in progress and that lots of things might and probably change place (toolbar, inspector controls etc..).

Having said that, I think it's great that there is not only design mock ups and iterations (that are definitely needed), but also feedback and suggestions! 💯 

One thing we probably could do is to get some decisions about some of the current options and implement them. For example if there seems to be a consensus on where the `search` input should live, we should do it and then iterate more.

**Comment by mapk** [comment]
Riffing off of @bph's comment above, here's a couple iterations of the Pagination icon toggle.

### Option 1 - Pagination
We can load the block with that setting "on" and highlighted. It would look like this:

![pagination icon 1c](https://user-images.githubusercontent.com/617986/93127314-8671ae80-f682-11ea-9cd4-aca4ff13e313.png)

### Option 2 - Pagination
Rather than a toggle, make use of the dropdown ability to switch the toolbar setting.

![pagination icon 2](https://user-images.githubusercontent.com/617986/93125915-a30ce700-f680-11ea-9ffe-95e77079ce1d.png)

**Comment by bph** [comment]
Could you have it just one line and use a toggle switch? 
<img width="255" alt="Screen Shot 2020-09-14 at 3 15 04 PM" src="https://user-images.githubusercontent.com/39980/93128377-5b488880-f69d-11ea-9fcb-9f0143dd87ac.png">

I know this is just an example. 

There is a Query Pagination block as well. Shouldn't there be a little note somewhere that it needs to be added to the space? Otherwise, the dependency is not obvious.

**Comment by mapk** [comment]
>Could you have it just one line and use a toggle switch?

Yes, that's how I had it in some of the [early iterations](https://github.com/WordPress/gutenberg/issues/25198#issuecomment-690506249). 

>There is a Query Pagination block as well. Shouldn't there be a little note somewhere that it needs to be added to the space? Otherwise, the dependency is not obvious.

That block would automatically be added if the Pagination is turned "on".

**Comment by mapk** [comment]
I'm thinking about the amount of input fields involved with the Query block settings. I've been working around creative ways (see above iterations) to show these form fields without breaking from existing UI patterns, or incorporating future patterns that are in development (Link UI). For the most part, block toolbars include simple toggle switches (ie. Bold, Italic) and dropdown selections (ie. alignment) whose states are reflected in the toolbar itself. There are a few that include more advanced controls like the Link and Color settings, but these are rare.

I'd like to include the more important settings in the toolbar, but because of the amount of form fields, I believe if we provide strong defaults, these settings _could_ remain in the sidebar where most of the form fields for blocks tend to reside. Unless there is an above approach that is more desired.

### Settings overview
![Screen Shot 2020-09-15 at 8 58 16 AM](https://user-images.githubusercontent.com/617986/93235180-0b67d100-f732-11ea-9e38-cf77e6cd84e1.png)

### Notes
- The Query block is expected to support various patterns/layouts. I have yet to explore how these integrate into the UI.
- This is the base-block that includes all settings and is meant to be a site-building block, not necessarily a common user-facing block.
- Block variations will be created that will be more simplified for common usage (ie. Latest Posts block).
- Yes, there are a lot of settings surfaced in the sidebar. Moving important ones to the toolbar means working in form fields into dropdowns or inline with the toolbar as displayed in previous mockups.
- To see the many other explorations, here's the [Figma file](https://www.figma.com/file/BdHpUk8pmEdKUdpEVL9eZv/Query-block-and-friends?node-id=58%3A0).

### Feedback
I'd like to get some theme developer's feedback on this. 
- Am I missing any settings that you feel should be included here?
- Which ones do you think should be moved to the block toolbar?

**Comment by aristath** [comment]
A few notes on that last iteration:

* Does the "include sticky items" exclude sitckies when not checked? Or does it simply return them to their normal position, unsticking them? If the latter (which would be correct) then the title would need changed.
* Offset should be right below the number of posts (items/page) instead of in the sorting section
* Perhaps inside filtering we can merge categories & terms/tags with the taxonomies? After all, that's what they are... That should help clean up the UI a bit if all taxonomy terms are under one roof

Critical & missing:
* Post-type

**Comment by mapk** [comment]
>Does the "include sticky items" exclude sitckies when not checked? Or does it simply return them to their normal position, unsticking them? If the latter (which would be correct) then the title would need changed.

It includes them when "on" and excludes them when "off". They would show as sticky in the block and remain at the top. cc @ntsekouras for confirmation.

>Offset should be right below the number of posts (items/page) instead of in the sorting section

Good eye! I've been pulling and pushing settings back and forth between the toolbar and sidebar. If they reside in the sidebar, I'll make sure that happens.

>Perhaps inside filtering we can merge categories & terms/tags with the taxonomies? After all, that's what they are... That should help clean up the UI a bit if all taxonomy terms are under one roof

I'm really unfamiliar with taxonomies, but are you imagining a dropdown that allows me to select between categories, tags, or taxonomies? Or just having them all lumped together?

**Comment by aristath** [comment]
> I'm really unfamiliar with taxonomies, but are you imagining a dropdown that allows me to select between categories, tags, or taxonomies? Or just having them all lumped together?

I'm imagining selectors similar to the ones we currently have for selecting tags. So not exactly dropdowns... Rather textfields with auto-complete functionality and a "dropdown selector" for lack of a better word once you start typing. This can work well for all taxonomies, allows selecting multiple items and is so compact that we could fit together Categories, Tags and any other custom taxonomies the post-type might have, we'll need 1 text-field/tag-selector per-taxonomy.
It's simple, functional and takes a lot less space than the categories selector. Since in the context of a query we don't need to add new terms, there really is no reason to show the full categories UI.

**Comment by kellychoffman** [comment]
The UI in the [latest design iteration](https://github.com/WordPress/gutenberg/issues/25198#issuecomment-692822019) to choose what to filter + display is that of -creating-. I think this is too confusing and the UI should use -filtering- patterns to make it more immediately recognizable as to what these settings are for. Doing so would have an added benefit of a more simplified and concise interface and perhaps could lead to a way to have some / all of this in the toolbar.

**Comment by mapk** [comment]
Thanks, @kellychoffman! Here's an iteration of filters looking more like filters. 

**NOTES**
- I researched how filters are used on other sites [in this Figma file](https://www.figma.com/file/BdHpUk8pmEdKUdpEVL9eZv/Query-block-and-friends?node-id=169%3A317) and they are primarily lists of checkboxes. Some include range controls and others as buttons.
- I went with checkboxes, but iterated on tags a bit to add some variation that related to how people often think of tags (ie. some sort of container with text in it).
- I added a section for "Applied filters" (still thinking about this). Normally filters are all displayed together somewhere easy for the user to see and remove if desired.

![Screen Shot 2020-09-15 at 5 13 07 PM](https://user-images.githubusercontent.com/617986/93277650-33c5ee80-f777-11ea-90e3-ad5549960214.png)

**Comment by mariohamann** [comment]
## Feedback

1. First of all, I think the sidebar is the right place for those options!
2. I think, if a tag (or "taxonomy without hierarchy") is selected, on should always be able to unselect within the same area. In the last exploration, "CMYK" and "dodge" don't look clickable and it seems one has to use the "applied filters" area to uncheck.
3. I'm not sure, if the overview of "Applied filters" in its current state is needed and suitable. To make it work, every "applied filter" should display what kind of filter it is, as e. g. "portfolio" could be a tag or category, too.

Attached my proposal for unchecking "taxonomies without hierarchy":

![image](https://user-images.githubusercontent.com/26542182/93299258-10418900-f7f5-11ea-9013-589ce39dc3e3.jpeg)

And to get a better overview of "Applied filters" I would suggest to use the taxonomy area again – here are two possibilites: 
showing the amount of selected taxonomies or a small circle, indicating "Inside, something is selected" (I think Ghost Blocks uses a similiar approach for e. g. individual margins and paddings):
![image](https://user-images.githubusercontent.com/26542182/93300063-68c55600-f7f6-11ea-8f90-ab8fbbca172b.jpeg)

## Some other thoughts

- As far as I understand, the query block should make it possible to show every Post Type – event Custom Post Types?   ( I think this should be the top selection, even before "display settings". At the moment I can't find the selection field. :)
- Will it be possible to list "Taxonomies", too (using WP_Tax_Query)?
- Will it be possible to add individual query arguments, e. g. date queries, meta field queries etc.? What do you think about adding a field, where one could insert queries in JSON format or PHP arrays? Or a button like "Add other query argument" showing a modal with every possible [core query argument](https://developer.wordpress.org/reference/classes/wp_query/)?

At least for the last one I think I should add a new issue. :)

**Comment by aristath** [comment]
I still feel that the tag-selector would be a more suitable UI element for this... This block will have way too many settings, so making the UI a bit shorter will help users have a better overview without having to continiously scroll.
And in case the "tag-selector" I previously mentioned wasn't clear, I meant this one:
![Peek 2020-09-16 09-50](https://user-images.githubusercontent.com/588688/93302250-581add00-f802-11ea-990a-1ab8e6c8b81d.gif)

It will allow us to significantly shrink the height of things and can be used for all taxonomies (categories/tags/others)

**Comment by mariohamann** [comment]
> I still feel that the tag-selector would be a more suitable UI element for this...

You are absolutely right! Please ignore my last proposal. :)

**Comment by ntsekouras** [comment]
>That block would automatically be added if the Pagination is turned "on".

For start pagination should be available to be added in many places and that would happen from the user by adding the `QueryPagination` block. For example a user might want to add pagination both on top and on the bottom.. Having said that, there is the possibility that the `show pagination` would not be needed and controlled by wether the user adds the `pagination` block. I'll have to check it better in code though so take this lightly.

>Critical & missing: Post-type

Yes, this is to be implemented in `Query` block.

>Perhaps inside filtering we can merge categories & terms/tags with the taxonomies? After all, that's what they are... That should help clean up the UI a bit if all taxonomy terms are under one roof

This is indeed an option and something that will become more clear when support for `taxonomy_relation` and `taxonomies` will be implemented. This is a hard one because it would be really difficult (IMO) to make it understandable. For example if you we have only one input about taxonomies and want to filter by a `tag` and a `category`, how these would show inside the input? Like a term with a label `green:tag, news:category`? And what about then if we implement the exclude, that might probably want to. That would mean two `taxonomy` inputs, include and exclude.

>we could fit together Categories, Tags and any other custom taxonomies the post-type might have, we'll need 1 text-field/tag-selector per-taxonomy.
This could be so overwhelming with many `taxonomies` and I don't believe it would work..

A solution might be here something like a dynamic (add/remove) list of taxonomy filters that would look like:
> selectbox: with taxonomies || selectbox: include/exclude || terms input: like is now allowing multiple terms

I hope this makes sense and if not please tell me to clarify.

>I still feel that the tag-selector would be a more suitable UI element for this... This block will have way too many settings, so making the UI a bit shorter will help users have a better overview without having to continiously scroll.

I totally agree with this.

**Comment by aristath** [comment]
> This is indeed an option and something that will become more clear when support for taxonomy_relation and taxonomies will be implemented. This is a hard one because it would be really difficult (IMO) to make it understandable. For example if you we have only one input about taxonomies and want to filter by a tag and a category, how these would show inside the input? Like a term with a label green:tag, news:category? And what about then if we implement the exclude, that might probably want to. That would mean two taxonomy inputs, include and exclude.

I apologise for not making what I meant more clear.
I meant We should merge them under the same section, not the same control. So they would still be separate selectors, but they would all be like the tag-selector to save some space.
Having them all in a single selector would be pretty messy and shouldn't be done. But instead of having filters ordered like
* Categories
* Tags
* Authors
* Taxonomies

it should instead be
* Categories
* Tags
* Product Brand (or whatever taxonomy name we have)
* Author

or even 
* Taxonomies
	* Categories
	* Tags
	* Product Brand (or whatever taxonomy name we have)
* Author

**Comment by mariohamann** [comment]
> * Taxonomies
>    * Categories
>   * Tags
>   * Product Brand (or whatever taxonomy name we have)
> * Author

I like this best and – based on my experience – it fits the approach of creating WP_Queries in PHP.

**Comment by aristath** [comment]
> I like this best and – based on my experience – it fits the approach of creating WP_Queries in PHP.

It does, yes - and it's very dev-friendly. But it's not as user-friendly... A normal user doesn't know what a "taxonomy" is, they just want to add a category :laughing: It's a thin line we need to walk.

**Comment by mariohamann** [comment]
> It does, yes - and it's very dev-friendly. But it's not as user-friendly... A normal user doesn't know what a "taxonomy" is, they just want to add a category 😆 It's a thin line we need to walk.

Hm, that's right. But would you say, someone using the "Query Block" and its deeper settings is a "normal user"?

**Comment by mariohamann** [comment]
In my opinion it is very important to split "Order" and "Orderby":
- `Order` should only contain "Ascending ▲" and "Descending ▼"
- `Orderby` should at least contain "title", "date", "modified", "comment_count" and even "random". Personally I would love to see "meta_value" and every other possible option described in [Codex](https://developer.wordpress.org/reference/classes/wp_query/#order-orderby-parameters) as well. :) By the way: Theoretically it is possible to add an array of `orderby` parameters... but maybe that's something for later versions. :)

**Comment by aristath** [comment]
Regarding `random` order, it's a tricky thing and I don't know if it should be available by default... Various hosts (including WPEngine) [disable random ordering](https://wpengine.co.uk/support/about-order-by-rand/) so perhaps that should be opt-in via a plugin. Or even opt-out via a PHP constant so hosts can just disable it if that is their policy.

**Comment by mariohamann** [comment]
> Regarding `random` order, it's a tricky thing and I don't know if it should be available by default... Various hosts (including WPEngine) [disable random ordering](https://wpengine.co.uk/support/about-order-by-rand/) so perhaps that should be opt-in via a plugin. Or even opt-out via a PHP constant so hosts can just disable it if that is their policy.

Interesting! I have to admit that I never used this parameter, but I thought that it may be interesting for some contexts. Thank you for this worthful clarification!

**Comment by ntsekouras** [comment]
>But would you say, someone using the "Query Block" and its deeper settings is a "normal user"?

It probably will be used more from theme developers and the goal is to have some block variations like `latest posts` to be more user-friendly for common cases.

>A solution might be here something like a dynamic (add/remove) list of taxonomy filters that would look like:
selectbox: with taxonomies || selectbox: include/exclude || terms input: like is now allowing multiple terms

I just wanted to drop [this here again](https://github.com/WordPress/gutenberg/issues/25198#issuecomment-693298641) for some thoughts..

There are only partial support for sorting by WP REST API. `random` I know for a fact that will not be implemented and has been discussed in `core` a lot, for performance issues. I have raised the issue about the supported options in core slack here: https://wordpress.slack.com/archives/C02RQC26G/p1598860035004400

**Comment by fklein-lu** [comment]
Assuming that the Query block mirrors the `WP_Query` class, a challenge that I see is whether users understand that these are all _and_ filters, not _or_.

To illustrate let's take this query:

```
$posts = new WP_Query( [
	'category_name' => 'foo',
	'tag' => 'bar',
] );

```

Quite often developers think that this retrieves posts that either have the category "foo" _or_ the tag "bar". But what it really does is get posts that both are in the "foo" category  _and_ have a "bar" tag. To have an _or_ query, you need to use a `tax_query`.

I'm not sure how we could have an interface that does reflect this. I remember Jetpack's Widget Visibility feature having similar challenges for determining on which pages widgets should be displayed.

There are also other behaviours that are easier to understand in code, but difficult to represent. For example I can have this query:

```
$posts = new WP_Query( [
	'category_name' => 'category__not_in',
] );

```

This will show all posts, except the ones with this particular category. I'm not really sure how a UI could represent that.

**Comment by mapk** [comment]
I'm catching up! Thanks for this back and forth discussion, everyone! ❤️ 

>I think, if a tag (or "taxonomy without hierarchy") is selected, on should always be able to unselect within the same area. In the last exploration, "CMYK" and "dodge" don't look clickable and it seems one has to use the "applied filters" area to uncheck.

@mariohamann, I hadn't shown the interaction yet, and probably should soon. The tags in that section would be clickable to deselect them again for sure. I agree with your request here. 😄 

>I'm not sure, if the overview of "Applied filters" in its current state is needed and suitable. To make it work, every "applied filter" should display what kind of filter it is, as e. g. "portfolio" could be a tag or category, too.

I think this is actually really important here. Because these are filters being applied to the results, getting it as close to a common filter pattern used throughout the web would be good. For example, take a look at this site, and notice that applied filters are highlighted at the top. Also notice they don't indicate what "type" of filter they are. It's a great way for the user to quickly see all the filters they've applied in one area.

![filters](https://user-images.githubusercontent.com/617986/93396764-05a2e600-f82d-11ea-9057-5ce7adfdfb84.png)


>And to get a better overview of "Applied filters" I would suggest to use the taxonomy area again – here are two possibilites:
showing the amount of selected taxonomies or a small circle, indicating "Inside, something is selected" (I think Ghost Blocks uses a similiar approach for e. g. individual margins and paddings):

I like those mockups you did! However, when the modules are all opened, they are still so spread out that it's difficult to tell just what filters are being used.

>I still feel that the tag-selector would be a more suitable UI element for this... This block will have way too many settings, so making the UI a bit shorter will help users have a better overview without having to continiously scroll.

@aristath, as a user, it's nice to see a list of items available to select from. I can't possibly remember all the tags on my site. For this reason, exposing this list somehow is important. However, it could be argued that when a user is building with this block, or a variation, they already have a in mind the specific filter they most likely want to use. 🤔  

I do think the tags could be iterated to resemble the Categories with a list of checkboxes possibly. I just went with a group of tags alphabetically because it resembled a "tag" feel and broke up the list of filters with lists. 😉 

>A solution might be here something like a dynamic (add/remove) list of taxonomy filters that would look like:
selectbox: with taxonomies || selectbox: include/exclude || terms input: like is now allowing multiple terms

@ntsekouras let's work through this together soon. I'd like to try this out, but need a few more details. Maybe we can hop on a zoom call and discuss it while I build it in Figma.

>This will show all posts, except the ones with this particular category. I'm not really sure how a UI could represent that.

@fklein-lu you're right. I've only scratched the surface of the "not include" UI. 

>a challenge that I see is whether users understand that these are all and filters, not or.

I agree. However interaction will reveal this pretty quickly I imagine. As the user makes filter changes, the list updates in the editor providing the user real time feedback for how the settings are being applied.

**Comment by mariohamann** [comment]
I would love to see something like this:

![image](https://user-images.githubusercontent.com/26542182/93453410-f29a1f80-f8d9-11ea-96ea-587716209190.jpeg)

Especially with some MAGIC-VARIABLES like `%today%`,  `%current_year%` etc., combined with a good documentation and examples I think this would be just great! (And maybe it helps to make the UI of the block lighter and to focus on the most important elements, but still making the block very flexible).

**Comment by aristath** [comment]
> I would love to see something like this:

The "custom arguments" thing would actually be pretty amazing... And could possibly help with https://github.com/WordPress/gutenberg/issues/25377 too :+1:

**Comment by ntsekouras** [comment]
>I would love to see something like this:

I think this would cause more problems that it would solve:

1. This seems like an option for super super users :) It makes the feel of this block much more complicated.
2. The REST API doesn't support many things from `WP_Query` and that could result to errors that users wouldn't understand why it's not working. They could try many params and might lead to even frustration? Personal thoughts here..

Tax Query needs to be implemented and will, but needs to do it under the hood and hide that complexity.

**Comment by mapk** [comment]
Taking in more of this feedback, I've worked on an incremental change.

- Added Post Types
- Added Search Terms
- Iterated on the Tags UI

![Screen Shot 2020-09-21 at 12 37 28 PM](https://user-images.githubusercontent.com/617986/93813214-97cd3480-fc07-11ea-871b-ccebaa9daffe.png)

**Comment by mapk** [comment]
### Applied filters 
The applied filters act as a single area to identify all filters being applied to the query.

![Screen Shot 2020-09-21 at 12 52 27 PM](https://user-images.githubusercontent.com/617986/93814503-766d4800-fc09-11ea-989a-d24e12ecbb67.png)

### Feedback
- Should we show a msg when no filters have been applied as in the example above?
- Should we just hide that accordion module if there are no filters?

**Comment by munirkamal** [comment]
Hey @mapk 

This is really coming out nice. Just a little feedback on your latest version.

I suppose the first thing to choose in the query block is the 'Post Type' then based on the selected post type you need to present the relevant 'Taxonomie' to choose. And then, finally, display the option to select specific 'Taxonomy Terms' based on the chosen 'Taxonomie'.

Not, sure if you have this already considered. The current design displays options to choose 'categories', 'tags', and even custom 'taxonomies' irrespective of the post type.

Hope you get my point.

**Comment by mariohamann** [comment]
Personally I'm a bit uncertain if the current way matches real needs and expectations. Therefore I would like to provide different use cases, to see if the UI fits them. I don't want to bring in further complications, but I think this block will be very important and therefore it should be build on a solid and scalable foundation.

## Use Case 1: FSE – Blog-Archive
### Description
Backend user want to build a shiny theme. Therefore s:he edits the blog archive and would like to make the newest post bigger (as e. g. on [WP Tavern](https://wptavern.com)). Of course on the second page the first item should be as big as the other posts.

![Screen Shot 2020-09-22 at 09 58 20](https://user-images.githubusercontent.com/26542182/93857002-3023f180-fcba-11ea-845e-dbc28fc5a4f0.jpg)


### Expected behaviour
The user visits the archive.php-settings and iserts two Query blocks. (Is this right? In fact that would be mine and one of mine @aristath 's https://github.com/WordPress/gutenberg/issues/25462#issuecomment-695978346 expected behaviour :) )
- Block 1:  `Items per Page: 1` and `Pagination: OFF`. Everything else should be default or inherited.
- Block 2: `Items per Page: 9`, `Pagination: ON`, `Offset: 1`. Everything else should be default or inherited.

### Questions arising
- Should there be options like `Include/Exclude/Only`? (e. g. if someone wants to show only sticky posts)
- Is it ensured, that there will be only one "Pagination"-block that uses the arguments from Block 2 (especially the offset)? Maybe "Used for pagination" would be better? Or should the pagination block autoamtically select the last query block? Or you could select a query block in the pagination block?
- Will those settings be used for taxonomy/author archives as well? Does the user have any possibility to change the behaviour, e. g. that the first post shouldn't be bigger as the others on those specific archives?
- Will filter options be locked or invisible in FSE? I see a lot of problems coming if users select taxonomy filters as they would break the complete logic of specific archives.
- Of course there should be no option to select different Custom Post Types in FSE – I think?
- Therefore: Should all those settings be in scope of FSE and Query Block or are those settings more a "Theme thing" – and the Query Block inherits EVERY argument?

## Use Case 2: FSE – Archive of Custom Post Type "Person"

![Screen Shot 2020-09-22 at 09 58 24](https://user-images.githubusercontent.com/26542182/93857007-34e8a580-fcba-11ea-8236-0cb979acd1b6.jpg)

### Description
The user wants to style the appearance of the Custom Post Type "Person" which is provided by a plugin. The plugin provides some custom taxonomies e. g. "Department" and "Position" (e. g. Trainee, Employee, External). S:he wants to show every person and order them from A–Z, overwriting the plugin's/WP default being "Publishing Date Descending".

###  Expected behaviour
The user visits the person-archive.php-settings and inserts one Query block. S:he changes the settings to `orderby: name` and `order: ascending"`. Furthermore s:he sets `Items per page: 0` to get every person.

### Questions arising
- In addition to the questions above I ask myself: If `Items per page: 0` is set, will the option to paginate be invisible as it has no effect? Will there be a warning that this could break the page while loading if there are a lot of persons to display?
- The Custom Post Type has no option of "Sticky Posts" – will this option be invisible?
- Will there be an option reset the settings, taking default from plugin? And: Do have plugin authors the option to set the "default" state for the query block? (In this case A–Z make a lot more sense.)
- Again I think that there should be no option to select specific taxonomies to prevent the bevhiour of specific archives.

## Use Case 3: Page Editor – Blog posts in "Hall of Fame" of specific tag
### Description
Now it's getting interesting. We are leaving FSE – the backend user wants to dynamically provide the most discussed posts of the Tag "UX" on a static page. 
![Screen Shot 2020-09-22 at 09 50 22](https://user-images.githubusercontent.com/26542182/93856362-39f92500-fcb9-11ea-9d25-b9079b46d6a3.jpg)


### Expected behaviour
The user adds a new page, writes some text and adds the Query block. S:he – deviating from standard – selects `Post Type "Post"` (Okay, maybe this is default :)), `Items per Page: 3`, `orderby: comments`. Furthermore the Tag "UX" is selected.

### Questions arising
- The user exactly knows, what kind of tag s:he wants to select, as this selection is "content" driven. S:he always selects tags with the current tag selector – why should this behaviour change in this block?
- Should the option for `pagination` be hidden here? In my opinion there is no possibility this could work outside of FSE despite of implementing some AJAX functions – and I'm pretty sure this is no option at the moment... :)
- Make it one step harder: Show the most commented posts of 2018. (I would still love to see a JSON-inserter for deeper for such cases. ;))

## Use Case 4: Page Editor – Department page showing their "newest trainees"
### Description
The last and maybe most interesting one. A department wants to motivate people to apply for an internship. Therefore, they want to show their current trainees dynamically.

![Screen Shot 2020-09-22 at 09 50 25](https://user-images.githubusercontent.com/26542182/93856375-3ebdd900-fcb9-11ea-9fb1-683b7082b95f.jpg)


### Expected behaviour
The user adds a new page, writes some text and adds the Query block. S:he – deviating from standard – selects `Post Type "Person"`, `Items per Page: 0`, `orderby: date`, selecting `Department: "Design"` and `Position: "Trainee"`

### Questions arising
- The Post Type "Person" has no "Category" and no "Tags", therefore those two taxonomies shouldn't be shown?
- Instead the taxonomies "Department" and "Position" are needed. The first one being with, the second without hierarchy. Again: In my opinion this should match the current interfaces, being "Checklist" for hierarchical and "Combo box" for not-hierarchical.

## Interim conclusion
- The Query block is a complex one – I think it's settings need to be very dynamic depending on context (FSE, Page Editor) and selected Post Type.
- Personally I would try to match known UX interactions as best as possible, so the users immediatelly know: "Here I select my tags – like in the Post editor", "Here I can select my categories". Furthermore I'm not sure, if an unordered list of e. g. 100 tags without the possibility to search would be ideal (thinking of a WooCommerce shop I developed where the Shop manager added such a big amount of tags...), especially in terms of a11y (In fact I have to admit I don't know how accessible a combo box is...).
- In my opinion, the applied filters section is not needed.
  - If for unordered taxonomies a combo box is used, the user directly sees the selected the options. If we use the same list for ordered taxonomies, the selected categories will be at the top (mayber after saving – but that's the current behaviour when selecting posts).
  - Furthermore you would have to show, to which Taxonomy a single Term belongs ("is it Category? A Taxonomy? A Department? A position?) 
  - I think it's strange that you can uncheck such filters in an backend at two positions.
- What about the possibility to insert single items (e. g. with Post ID), as in the WooCommerece Product block?

The examples you have shown @mapk  are from shops and I think this is a complete different use case.
  - Shops want to show what they have, want to give possibilities. Shoppers use filters to look for specific items or to poke around. For frontend users, e. g. the tag-selection and the overview of applied filters as shown above would be great, because they switch those filters fast and often.
- Query block users in my opinion don't poke around, trying to find a specific block. They are content driven and they know their content. If the block works as expected, they select their options and are done. They don't do a lot switching back and forth – I think?

**Comment by mapk** [comment]
>I suppose the first thing to choose in the query block is the 'Post Type' then based on the selected post type you need to present the relevant 'Taxonomie' to choose.

@munirkamal, great point. I've noted for the next iteration that the Post type will directly impact the filter options. 👍  

@mariohamann I love those user flows! Thanks for wrangling that. You've got so many good points and questions that I'll try to address.

>The Query block is a complex one – I think it's settings need to be very dynamic depending on context (FSE, Page Editor) and selected Post Type.

Yes, I agree! This particular block is meant as a base block that would include all the settings that a block like this might need. For more targeted flows, as you've listed above, we'd most likely have a block variation of this one that would align better. For example, the Latest Posts block will be a variation of this one specifically focused on Posts, most likely sorted by publish date.

>Furthermore I'm not sure, if an unordered list of e. g. 100 tags without the possibility to search would be ideal

This could be solved by including a search field which I'm not opposed to. But I think adjusting the UI to mimic a filtering system is still relevant here.

>In my opinion, the applied filters section is not needed.

I understand your reasoning. Keep in mind there may be many screen sizes for which not all filters are visible on screen. Having them listed in one area helps alleviate this and provides a high level, easy review of what the block is filtering. I'm pushing for this because I believe its beneficial, but I'm open to opposition! If it proves unnecessary, we can easily pull it out.

>Furthermore you would have to show, to which Taxonomy a single Term belongs

This more meta information can be found in the specific filters below. Adding that information here could cause it to become more messy, but it's worth an exploration. I'll mock this up.

>I think you should add "Order" and "Order by" to display settings, too.

Do you think this is important enough to get into the first iteration of this block?

>Query block users in my opinion don't poke around, trying to find a specific block.

While they may not be looking for a specific block (except in your use case 1, block 1) they are looking for a specific list of content. Turning on/off filters to get that content right could be an exploration for them. I doubt they do a lot of switching filters once they're accustomed to how the filters change their results, but getting to that happy place might involve toggling several settings.

>As especially the filtering is very complex – maybe we should start with the display settings to provide the block for FSE and add the filter options later on?

I like that you're pushing for an MVP (minimum viable product). But I think the filtering is going to inform the block variations (and vice versa) as we continue. This means I'd like to nail down most of the filtering in this block right away since its providing the base to these other blocks. Please keep the feedback coming!

**Comment by ntsekouras** [comment]
Hey @mariohamann ! - Thanks for the detailed comment!

>The examples you have shown @mapk are from shops and I think this is a complete different use case.
>They are content driven and they know their content

>how these would show inside the input? Like a term with a label green:tag, news:category
I believe it too and arises more UI problems as I have mentioned above(https://github.com/WordPress/gutenberg/issues/25198#issuecomment-693298641) about how to distinguish the filters. A `tag` can have the same name with a `category` and a `custom taxonomy`.

>I think you should add "Order" and "Order by" to display settings, too.

Order settings are Query settings and not display settings.

>@ntsekouras let's work through this together soon. I'd like to try this out, but need a few more details. Maybe we can hop on a zoom call and discuss it while I build it in Figma.

I think @mapk we should explore the option I propose as well.

> As especially the filtering is very complex – maybe we should start with the display settings to provide the block for FSE and add the filter options later on?

Since there are many things to answer and consider, I think this a good direction. I have in mind to start adding some more display options to `PostBlocks` to enhance the current functionality/experience with `Query` block. Related PRs here:
https://github.com/WordPress/gutenberg/pull/25341 (merged)
https://github.com/WordPress/gutenberg/pull/25412 (needs review)

This doesn't mean to not progress Query block, but both being worked in parallel.

**Comment by kellychoffman** [comment]
Design feedback:
- Now that we have a handle on what settings to include, this might be a good time to see the design in the add and edit flow to make sure both are intuitive. (Meaning, seeing this in context will help inform and prove some of the decision decisions rather than continuing to view the sidebar and toolbar settings in isolation)
- Seeing what filters are applied is going to be very useful, but it needs to be more apparent that "Applied filters" applies to all filters below. Overall, the hierarchy of the sidebar could be a bit more clear.
- The UI for all the filtering components still needs some work. Feels inconsistent and some (specifically tags) will have issues working at scale. For example, tags are displayed inline but categories are listed block style. You can search categories but not authors, etc. I am not saying to make them all look exactly the same, but they feel a bit all over the place and its unclear why. Filtering should be a fairly quick process and we could apply some "don't make me think" here.

**Comment by mapk** [comment]
>The UI for all the filtering components still needs some work. Feels inconsistent and some (specifically tags) will have issues working at scale. For example, tags are displayed inline but categories are listed block style. You can search categories but not authors, etc. I am not saying to make them all look exactly the same, but they feel a bit all over the place and its unclear why. Filtering should be a fairly quick process and we could apply some "don't make me think" here.

Oh, @kellychoffman, I've got an answer for ya!

- Tags can still work at scale much like the Categories do. The module would only show so many so as not to take up the entire sidebar if there are hundreds. But there is the "See all" if someone wanted to see everything. Otherwise, similar to the category functionality now, they can be searched.
- The reason why Tags are displayed differently from Categories is because tags are a single-level taxonomy. You can't add tags to a parent tag. Categories can exist in hierarchical levels meaning that a category can exist under a parent category. This is why I maintained a visual difference. It also breaks up the monotony of checkbox lists.
- As with any taxonomy, if the total exceeds a certain number, a search is revealed to make finding them easier. Authors, in the case of the mockups, doesn't exceed that number so there is no search needed since they are all visible.

I hope that helps!

I'll iterate to get these settings into flows soon! Thanks for your feedback.

**Comment by ItsJonQ** [comment]
Haiii!!! Hope you all are well. I poked around some of the query block UI as it feels like it's within the realm of "Design Tools", which helps me better understand use cases.

### 🥇 First idea

First off, I fully recognize how complex something like this is. The combination of things is (theoretically) endless, compounded by the amount of categories, tags, custom taxonomies a user may have. (I've seen WP sites with literally 10,000+ tags).

That being said... I think there may be a UI solution here - at least to handle the combinations.

My mind jumps to a "**Condition Builder**" UI.

Example:
![advanced_condition_builder_02](https://user-images.githubusercontent.com/2322354/95586644-f7508000-0a0e-11eb-8c82-cd09177460b5.png)

This idea sparked from the following comments in the thread:

@ntsekouras 
A solution might be here something like a dynamic (add/remove) list of taxonomy filters that would look like:
selectbox: with taxonomies || selectbox: include/exclude || terms input: like is now allowing multiple terms

@mariohamann 
https://github.com/WordPress/gutenberg/issues/25198#issuecomment-694119676

### 🧬 Condition Builders

Condition builders are designed to handle this very thing. I may be biased as I've worked on a Condition Builder UI several years ago. It was, as you could imagine, wildly complex (as far as combinations of things goes). However, the nature of a condition builder was able to it.

A bonus is that folks may already be familiar with Condition Builder UIs, especially if they have experience in the analytics or marketing world where Condition Builders are standard (e.g. Email marketing platforms)

It is worth noting that Condition Builders are more of an "intermediate" feeling UI. Not as friendly as a couple of checkboxes. However, I think the Query block is "intermediate" by nature. Much more so compared to a "Paragraph",

### 🧠 Powerful logic

In addition to adding/chaining filters... operators can be added such as `include`, `exclude`, `and`, `or`. Each one compounding the flexibility of the query, while adding minimal UI. It's pretty amazing.

Something like... (pseudo code incoming)

```
* Include 'CATEGORY: TECH', 'TAG: APPLE'
* Exclude 'CATEGORY: REVIEW', 'TYPE: SPONSORSHIP'
* Post date within '3 years'
```

### 🌍 Query "Language"

The condition builder (query builder) is modelling after a "query language". In our case, WordPress's query language, which resembles MySQL a bit. An example to look into may be Jira's JQL (Jira Query Language):
https://www.atlassian.com/blog/jira-software/jql-the-most-flexible-way-to-search-jira-14

It's been a while since I looked into it deeply.. From what I remember, the UI is basically a condition builder. They also have a complimentary text field where you could write the query manually, not unlike the suggestion from @mariohamann.

I'm not suggestion we create our own query language.

Rather, I'm suggesting the UI solution for Query block should map to both the mental model of how users assume data can be queried and the API of the WP query parameters.

---

No need to run with this idea. You all have been at this much longer than I have! The idea of a condition builder may have already come up (hard to tell from the sea of comments).

Hopefully this helps though!

---

P.S. If a Condition Builder UI is what y'all think is best, I'd be happy to help prototype some ideas to try things out 🏗️ .

**Comment by ntsekouras** [comment]
Thanks @ItsJonQ! I would really love to see this idea being explored as well! I think it might suit us well!

**Comment by mariohamann** [comment]
I really like the idea of the `query block` being a flexible "Query builder". 

I tested a concept some days ago, to see how we could use those builded Queries for the REST API to get the full power of WP_Query: https://github.com/mariohamann/rest-query-endpoint


![](https://user-images.githubusercontent.com/26542182/94361955-3b12c380-00b8-11eb-9f7f-79b70d11453e.gif)


GIven that, variants could still create some "pre selections" or "default filters". (E. g. for a `Recent Comments` block or similar).

**Comment by mapk** [comment]
While @ItsJonQ is working toward a conditional builder, I thought to take a step back based on some of @ntsekouras' earlier endeavors. Currently WordPress.com has a Blog Post block that displays their filters in a simply search input that shows selections as labels. This follows similarly to the Tags module in WordPress.org right now.

This solution...
- Reduces the UI.
- Provides strong consistency throughout each filter.
- Is predictable and a pattern used already (somewhat).

I've also included the current iteration of the block's toolbar. Some of that [interaction can be seen here](https://github.com/WordPress/gutenberg/issues/25750#issuecomment-704468135).

![Screen Shot 2020-10-13 at 4 06 45 PM](https://user-images.githubusercontent.com/617986/95925562-f4d38a80-0d6e-11eb-8d43-cbee140223bd.png)

**Comment by mariohamann** [comment]
## Filters
I think it is a very good idea to stick to common behaviour with the filters → Really like the new explorations @mapk . 👍🏻 Building on @ItsJonQ proposal I had this quick idea I would like to bring in:
![Screen Recording 2020-10-14 at 09 38 01](https://user-images.githubusercontent.com/26542182/95958190-15e4bb80-0e01-11eb-8e4c-9f98585bf51a.gif)
Clicking on "Open condicitional builder" opens a modal, which influences the shown JSON. Maybe it's to "technical" thought but maybe it helps for further explorations. :)

## Sticky posts
### tl;dr
For sticky posts the current UI seems to behave as a boolean (include, show only?), but we need three options (include, exclude, show only). Furthermore we need the option to "ignore_stickiness". @ntsekouras What are the planned options from code perspective?

### Explanation
In the REST API (referring the [docs](https://developer.wordpress.org/rest-api/reference/posts/#arguments)) we have the parameter `sticky` [boolean], saying "Limit result set to items that are sticky." Therefore we have 3 options.
- ` `: If not set, sticky and non-sticky posts show up. (Default)
- `sticky=true`: include **only** sticky posts (corresponding to `'post__in' => get_option( 'sticky_posts' )` in WP_Query)
- `sticky=false`: exclude sticky posts (corresponding to `'post__not_in' => get_option( 'sticky_posts' )` in WP_Query)

_🔍 Example "Featured posts" and "Other posts": If one wants to list all featured (=sticky) posts in one `Query` e. g. with big pictures etc. and all other posts (= non sticky) in another `Query` without images, we need the option to include and exclude. If we only have the option to include only sticky posts, they would show up in both `Queries`._ 

Fourthermore in WP_Query we have the parameter to `ignore_sticky_posts`. This ignores the "stickiness" of posts. If set, sticky posts still are shown but not at the top of the output. I couldn't find this option in the REST docs and didn't have the time to check the source code. But this option is very important.

_🔍 Example "Recent posts": Sticky posts should show up, but they shouldn't stay on top._

## Toolbar vs. Sidebar
### tl;dr
Because of other blocks/plugins with similiar behaviour (e. g. Core/Recent posts, WP.Com/Newspack/BlogPosts, Elementor) the complexity (e. g. two steppers and descriptions inside pagination) and power (e. g. Post Type selector) of the paramaters, I think the options of the toolbar should stay in the sidebar.

### New UI-Interactions
At the moment, I have the feeling there are way to many new UI interactions:
- "Toolbar option changes sidebar content.": The post type selector is responsible for the taxonomies showing up in the filter section. In my opinion this is lots of power for such a small button. :)
- "Inserters and steppers inside a toolbar option": Inside the pagination option there are steppers for pagination and offset. I had a look on some of the core blocks and could only find two examples, where options in the toolbar have such complex behaviour – but both are directly content specific: The color-selector in navigation toolbar and the URL selector e. g. in paragraphs or similar. Every other toolbar option have just some single options.

### Complexity
Often toolbar settings are used for visual changes (e. g. bold, italic, text-align, image position) or for simple actions (add new row in tables, add link etc.). In my opinion many of those options are far away from being self-explanatory and simple, e. g. sticky posts (see above), offset, pagination or the technical term "post types". To support the users, I think they definetely need good and verbal descriptions. Furthermore, I really like the icons in the toolbar, but they are very tiny and e. g. the `10/0` icon is not self-explanatory enough. I think verbal descriptions and labels would fit the complexity of this block a lot more than icons – and therefore I think the sidebar is perfect.

### Comparison to other blocks/plugins
I had a look on the `Recent Posts Block`, the `Newspack Posts Block` (thanks @mapk for bringing this in!) and Elementor `Posts widget`.  All of them e. g. have the option for changing the amount of posts showing up, in Elementor you even have the possibility to select post types. All of them have those options in the sidebar. I think it is for good reason: You need way less clicks to change those settings and you can directly see all options at once.

![image](https://user-images.githubusercontent.com/26542182/95956830-26943200-0dff-11eb-8c4b-6b3bfac9bc8d.jpeg)

**Comment by ItsJonQ** [comment]
Perhaps I may be overthinking things, but I think the designs may not accommodate more complex WP setups.
(But maybe those folks aren't the target users?)

What I mean by that would be folks with several custom taxonomies beyond Categories and Tags.
I think it gets trickier when taxonomy values are combined, like:

`category: Apple`, `tag:phone`, `tag:budget`, `media_type:video`, `editorial_type:review`

(just making up values above)

With the sidebar UI, we'd need to have 1 input per taxonomy.
(Not sure how it would work in the Toolbar, especially if values need to be combined).

That being said, I'm not opposed to anything. Just thoughts and observations from my side :)

---

I have another question! Would the Query block support [custom fields](https://wordpress.org/support/article/custom-fields/)? 😱 

If so, that makes it trickier considering the keys and values can be anything 🙈 

(Not impossible. The UI just needs to accommodate the flexible nature of it :))

**Comment by mapk** [comment]
Thanks, @mariohamann! When I was exploring more robust filters, mixing the filters in the sidebar with the display settings was overwhelming. Now with more simplified filters, it might be okay to bring some non-essential settings back to the sidebar like sticky posts. It's good to keep these [Gutenberg "best practices"](https://developer.wordpress.org/block-editor/designers/block-design/) in mind for informing the design.

>The primary interface for a block is the content area of the block
The Block Toolbar is a secondary place for required options & controls
The Settings Sidebar should only be used for advanced, tertiary controls

So with that in mind...

>"Toolbar option changes sidebar content.": The post type selector is responsible for the taxonomies showing up in the filter section. In my opinion this is lots of power for such a small button. :)

I think the Post Type is an important setting as well! It determines a lot of the other settings that can be used. Does moving it to the sidebar make it less important? Keeping it in the toolbar does give it more importance, but it's also a setting that once set is probably not often returned to.

>To support the users, I think they definetely need good and verbal descriptions.

We might be able to provide descriptions in the dropdowns if needed. 

>Furthermore, I really like the icons in the toolbar, but they are very tiny and e. g. the 10/0 icon is not self-explanatory enough. I think verbal descriptions and labels would fit the complexity of this block a lot more than icons – and therefore I think the sidebar is perfect.

Icons aren't great at expressing changes and reflecting settings especially if they are new icons that users aren't familiar with. Moving it all to the sidebar seems to make the sidebar overwhelming and lessen importance of many values. Maybe post types should just be words as you suggested?

@ItsJonQ, thanks for chiming in!

>With the sidebar UI, we'd need to have 1 input per taxonomy.

This is correct. I'm not all that familiar with taxonomies, or how often people have multiple taxonomies to choose from. Is this a common thing that could wind up showing 5-10 different taxonomies?

>For the Toolbar, I notice that it's showing custom icons for the Post Types.
I haven't looked into it, but I'm not sure how compatible the data from the REST API is with the editor (as far as icons are concerned, especially custom ones).

This is correct. @ntsekouras has raised this issue as well. It's where I'd like to get to, but if it can't be achieved with the API, it will need to be displayed differently. Maybe just using words here instead of icons as @mariohamann suggested might work well.

>I have another question! Would the Query block support custom fields? 😱

Great question. I not sure myself.

**Comment by mariohamann** [comment]
@ItsJonQ I think this block is perfect to overthink things (and needs it). 😄 

> With the sidebar UI, we'd need to have 1 input per taxonomy.
Two inputs. We need to seperate "Include" and "exclude" (e. g. including `category:apple` and excluding `category:restricted`). Here is an old draft we discussed.

![Screen Recording 2020-09-28 at 11 19 07](https://user-images.githubusercontent.com/26542182/96019834-641fac00-0e4d-11eb-97d0-7e13fc4f8fe0.gif)

I think the way @mapk shows above showing the main taxonomies (cateogry and tags) may be a very good one (as it may meet the expectations of about 98% of blogger...). And we could add "Advanced filters" with the Query Builder for the sophisticated rest of us. :) This would be inline with [WP_Query](https://developer.wordpress.org/reference/classes/wp_query/) where you can access those main taxonomies per default with `cat` ([link](https://developer.wordpress.org/reference/classes/wp_query/#category-parameters)) and `tag` ([link](https://developer.wordpress.org/reference/classes/wp_query/#tag-parameters))– and for the rest you need a more complex `tax_query` ([link](https://developer.wordpress.org/reference/classes/wp_query/#taxonomy-parameters)).

Custom fields: This would be just lovely. I think REST API can already handle them.

**Comment by annezazu** [comment]
[As part of the fifth call for testing for the FSE](https://make.wordpress.org/test/2021/04/14/fse-program-testing-call-5-query-quest/#comment-1501), the experience around what lives in the toolbar vs the block's sidebar was mentioned as being quite confusing: 

>Having some query controls in the block toolbar and others in the block’s sidebar seemed confusing. Colocating them would seem more logical. I appreciate we’ve probably placed the “most common” controls in the toolbar for convenience but having to jump between locations when customising the query didn’t make for a smooth experience. Perhaps duplicate the toolbar controls into the sidebar?

EDITED TO ADD: This came up during the [seventh call for testing as well as a point of confusion](https://make.wordpress.org/test/2021/05/26/fse-program-testing-call-7-polished-portfolios/#comment-1707):

>There’s a bit of a confusion point in the Query Block with Items per Page. Despite having multiple published posts only one appeared by default. I found the controls in the Block Toolbar to increase, but also found it a bit cumbersome to toggle between the Block Toolbar and Block Sidebar to refine the underlying query. 

Currently, there is a lot of needing to jump around to both sets of settings. For example, you might want the block to display a certain category but only 3 posts from that category. To do that, you have to interact with the block sidebar settings first to set the category before then using the block toolbar to select the number of visible posts.

**Comment by mariohamann** [comment]
> Currently, there is a lot of needing to jump around to both sets of settings. For example, you might want the block to display a certain category but only 3 posts from that category. To do that, you have to interact with the block sidebar settings first to set the category before then using the block toolbar to select the number of visible posts. 

As I reasoned in detail in https://github.com/WordPress/gutenberg/issues/25198#issuecomment-708222120 I still think it's much more intuitive to put every setting into the sidebar. Querying is complex, the block is complex, therefore hiding settings behind unfamiliar icons and distributing them across different places reduces usability.

(That's why I didn't use the toolbar in my prototypes.)

**Comment by annezazu** [comment]
More feedback around this experience has come up [in the tenth call for testing for the FSE Outreach Program](https://make.wordpress.org/test/2021/09/29/fse-program-testing-call-10-pattern-party/#comment-2131). This seems to be a repeated point of confusion:

>Query Loop Block: this time I was able to set the number of posts shown in the query loop block as I knew that this setting was in the block’s contextual toolbar, yet I’d expect this feature to show on the main block sidebar, as it might be one of the most common things user look for, in my opinion.

**Comment by ntsekouras** [comment]
@jameskoster any design input for this? 🙇

**Comment by jasmussen** [comment]
There will likely be cases in the future where we have a particular control available both in the inspector, and in the block toolbar. I'm thinking for example background color options for the Cover block. Given that as a pattern, it could make sense to visualize what it might look like if a query options panel held these options, with the same contents of the panel duplicated in the block toolbar flyout.

**Comment by jameskoster** [comment]
Swimming the sea of Design Tools, I can't shake the idea that we might use that UX to create a query builder of sorts:

https://jameskoster.design/2021/10/07/concept-creating-a-query-builder-inspired-by-design-tools/

I know this is a much bigger thing, but I think we should be careful about arbitrarily moving things around without a clearer vision for where we want this block (and it's variations) to go. That said – and to Joen's point – if we wanted to put these settings in the Inspector _in addition_ to the toolbar, I think that would be fine.

Edit: I opened a discussion for the query builder idea: https://github.com/WordPress/gutenberg/discussions/35760

**Comment by davewhitley** [comment]
To echo the past few comments on this post:

> if we wanted to put these settings in the Inspector in addition to the toolbar, I think that would be fine.

> Query Loop Block: this time I was able to set the number of posts shown in the query loop block as I knew that this setting was in the block’s contextual toolbar, yet I’d expect this feature to show on the main block sidebar, as it might be one of the most common things user look for, in my opinion.

Absolutely! It would be great to explore an advanced query builder, but I think the QL controls in the toolbar should at least be duplicated in the sidebar. That's where I first looked for those settings, and I didn't find them until several days later. I can imagine users might not ever find those controls if they aren't looking for them in the block toolbar.

I very much think of the block toolbar as an inline settings thing, like text appearance, paragraph alignment, width alignment, etc.

**Comment by ntsekouras** [comment]
Query Loop controls have change much since this issue was created. Let's close it.
